### PR TITLE
bug(snapshot) : Do not preempt inside OnDbChange issue #829

### PR DESCRIPTION
**Pull request by adiholden**
1. Serializer encode db index of entries
2. Do not encode db index inside consume channel
3. Allow push serialized data to channel from main snapshot loop and from journal callback (push may preempt)

### CODE REVIEW

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -681,6 +686,7 @@ error_code RdbSerializer::FlushToSink(io::Sink* s) {
   // interrupt point.
   RETURN_ON_ERR(s->Write(bytes));
   mem_buf_.ConsumeInput(bytes.size());
+  last_entry_db_index_ = kInvalidDbId;  // After every flash we should write the DB index again.
```
Comment (dranikpg) [inline src/server/rdb_save.cc:689]: flush

but why actually?

File: src/server/rdb_save.cc [hunk 2]
```diff
@@ -701,17 +707,20 @@ io::Bytes RdbSerializer::PrepareFlush() {
...
-  RETURN_ON_ERR(SaveLen(entries.size()));
+  RETURN_ON_ERR(SaveLen(1));
   RETURN_ON_ERR(SaveString(io::View(journal_mem_buf_.InputBuffer())));
-
+  // SELECTDB is serialized inside journal writer. In rdb loader when parsing journal blob
+  // the journal reader parses the entry db index, and we dont use the main flow of
+  // updating the current db index. Unless we change the flow in the loader, we must
+  // update last_entry_db_index_ to invalid so that the next snapshot entry to be serialized
+  // will update the db index.
+  last_entry_db_index_ = kInvalidDbId;
```
Comment (dranikpg) [inline src/server/rdb_save.cc:723]: So the journal db numbering and rdb db numbering are fully independent, right? Then it seems like we don't have to reset last_entry_db_index_?

File: src/server/snapshot.cc [hunk 3]
```diff
@@ -159,8 +159,8 @@ void SliceSnapshot::IterateBucketsFb(const Cancellation* cll) {
 
   // TODO: investigate why a single byte gets stuck and does not arrive to replica
   for (unsigned i = 10; i > 1; i--)
-    CHECK(!default_serializer_->SendFullSyncCut());
-  FlushDefaultBuffer(true);
+    CHECK(!serializer_->SendFullSyncCut());
+  PushSerializedToChannel(true);
```
Comment (dranikpg) [inline src/server/snapshot.cc:163]: 😅

**Review by dranikpg — COMMENTED** [review]
Lets use the new helpers from helio `EnterFiberAtomicSection` and `FiberAtomicGuard` in the snapshotting loop that iterates over other snapshots to assert no suspensions are made

File: src/server/snapshot.cc (hunk 3 above)
Comment (romange) [reply src/server/snapshot.cc:163]: OMG, maybe we do not? I suggest removing this loop and see if the issue remains.

File: src/server/rdb_save.cc [hunk 4]
```diff
@@ -263,7 +266,9 @@ error_code RdbSerializer::SelectDb(uint32_t dbid) {
 
 // Called by snapshot
 io::Result<uint8_t> RdbSerializer::SaveEntry(const PrimeKey& pk, const PrimeValue& pv,
-                                             uint64_t expire_ms) {
+                                             uint64_t expire_ms, DbIndex dbid) {
+  SelectDb(dbid);
+  last_entry_db_index_ = dbid;
```
Comment (romange) [inline src/server/rdb_save.cc:271]: is SelectDb used anywhere else?  
anyway, I think this line should be moved to  SelectDb -line 261

File: src/server/rdb_save.cc (hunk 1 above)
Comment (romange) [reply src/server/rdb_save.cc:689]: flash -> flush. (flash is 📸 ).

File: src/server/rdb_save.cc [hunk 5]
```diff
@@ -701,17 +707,20 @@ io::Bytes RdbSerializer::PrepareFlush() {
   return mem_buf_.InputBuffer();
 }
 
-error_code RdbSerializer::WriteJournalEntries(absl::Span<const journal::Entry> entries) {
+error_code RdbSerializer::WriteJournalEntries(const journal::Entry& entry) {
```
Comment (romange) [inline src/server/rdb_save.cc:710]: now it becomes singular `WriteJournalEntry`

File: src/server/snapshot.h [hunk 6]
```diff
@@ -107,20 +101,16 @@ class SliceSnapshot {
   // Close dest channel if not closed yet.
   void CloseRecordChannel();
 
-  // Call PushFileToChannel on default buffer if needed.
-  // Flush regradless of size if force is true.
-  // Return if flushed.
-  bool FlushDefaultBuffer(bool force);
+  // Push serializer's internal buffer to channel.
+  // Push regradless of buffer size if force is true.
```
Comment (romange) [inline src/server/snapshot.h:105]: regradless -> regardless

File: src/server/rdb_save.cc (hunk 4 above)
Comment (adiholden) [reply src/server/rdb_save.cc:271]: true this is left overs from my implementation before some fix, now its not relevant. will move inside

File: src/server/snapshot.cc (hunk 3 above)
Comment (adiholden) [reply src/server/snapshot.cc:163]: We still have bug here

**Review by dranikpg — APPROVED** [review]

**Comment by dranikpg** [comment]
You didn't add the new FiberAtomicGuard, but we can do it separately when pulling new helio

**Pull request merged**